ci/scripts/cpp_build.sh

```diff
@@ -287,6 +287,11 @@ fi
 if [ "${ARROW_USE_MESON:-OFF}" = "ON" ]; then
   time meson compile -j ${ARROW_BUILD_PARALLEL}
   meson install
+  # Remove all added files in cpp/subprojects/ because they may have
+  # unreadable permissions on Docker host.
+  pushd "${source_dir}"
+  meson subprojects purge --confirm --include-cache
+  popd
```

Comment:
@WillAyd This is a workaround.

Reply:
Hmm... This isn't a workaround... CI failed with this...

Reply:
Yes. If we don't have this, `cpp/subprojects/` has unreadable files. It causes an actions/cache error.

(I think that Meson should not change the source directory.)

Reply:
Hmm not sure I understand - Meson strictly forbids changes to the source directory. 

Not a blocker for now though - can be cleaned up later if needed

Reply:
This was executed on `ubuntu:24.04` Docker image with `apt update && apt install -y git gcc ninja-build meson cmake`.

Reply:
Ah OK - I was excluding subprojects from my consideration of the source directory, as Meson does unpack third party dependencies to that directory during the configure stage (i.e. before the build directory even exists)